### PR TITLE
Fix-12964: Set :format-rows false to export card results

### LINKED ISSUE #12964: Timestamp is always present in dates when exporting to CSV
**Describe the bug**
Dates are displayed without timestamp in visualization for native queries, but when exported they include timestamp.

**To Reproduce**
1. Create native query with a date
2. Set visualization to table
3. Go to settings and click "Off" in "Show the time"
4. (If you export the CSV before saving the question it works, strangely)
5. Save question
6. Export CSV
7. Open CSV and see that timestamp is still present

**Expected behavior**
I expect the CSV export to look just like what I am seeing in the table visualization, without the timestamp.

**Screenshots**
image

CSV:
```
date
2020-07-21T00:00:00Z
```

**Information about your Metabase Installation:**

```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.7+10",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.7",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.7+10",
    "os.name": "Linux",
    "os.version": "4.19.123-coreos",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mongo",
      "mysql",
      "redshift",
      "postgres",
      "athena",
      "h2"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.8"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.8"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2020-05-28",
      "tag": "v0.35.4",
      "branch": "release-0.35.x",
      "hash": "b3080fa"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

**Severity**
We use CSV exports from Metabase to load into Google Sheets. This is breaking the formatting of some of the sheets and formulas.

**Additional context**
Note: this happened when updated Metabase from 0.34 to 0.35. The DB this is happening on is AWS Redshift.

### PR DESCRIPTION
**Pull request by amommendes**
Fixes #12964 

The issue is related to the default value of `:format-rows` option.
In the `/:card-id/query/:export-format` [endpoint](https://github.com/metabase/metabase/blob/b6ad9f508971017f92f552d90ca9e7787272acc9/src/metabase/api/card.clj#L642) (which is used to export saved questios) we dont pass `:format-rows? false` as we do in the `/api/dataset/:export-format` [endpoint](https://github.com/metabase/metabase/blob/b6ad9f508971017f92f552d90ca9e7787272acc9/src/metabase/api/dataset.clj#L91). Because we don't pass `false` to `:format-rows` in card export endpoint, it'll assume the default value in the `format_rows.clj`, which is defined as  `true` in the `format-rows` [method](https://github.com/metabase/metabase/blob/b6ad9f508971017f92f552d90ca9e7787272acc9/src/metabase/query_processor/middleware/format_rows.clj#L77).
[Following the MBQL schema](https://github.com/metabase/metabase/blob/b6ad9f508971017f92f552d90ca9e7787272acc9/backend/mbql/src/metabase/mbql/schema.clj#L955), the default value of this parameter should be `false`. We prefer to change only the endpoint, because the consequences is more predictable. We also changed the docstring in MBQL schema to maintain the default value documentation consistent with the code.

### CODE REVIEW

File: src/metabase/api/card.clj [hunk 1]
```diff
@@ -639,7 +639,7 @@
      :parameters  (json/parse-string parameters keyword)
      :constraints nil
      :context     (dataset-api/export-format->context export-format)
-     :middleware  {:skip-results-metadata? true})))
+     :middleware  {:skip-results-metadata? true :format-rows? false})))
```
Comment (sbelak) [inline src/metabase/api/card.clj:642]: Can you just change indentation to our coding standard: each keyval in it's own row, values vertically aligned:
```
{:skip-results-metadata? true 
 :format-rows?           false}
```

**Review by sbelak — APPROVED** [review]
Looks reasonable. Thanks for the PR! A test for this would be nice though.

File: src/metabase/api/card.clj (hunk 1 above)
Comment (amommendes) [reply src/metabase/api/card.clj:642]: Sure

**Comment by sbelak** [comment]
p.s. If you need inspiration for the tests: https://github.com/metabase/metabase/blob/master/test/metabase/api/card_test.clj#L1102 

Of course I'm happy to help if you get stuck somewhere.

**Comment by amommendes** [comment]
> p.s. If you need inspiration for the tests: https://github.com/metabase/metabase/blob/master/test/metabase/api/card_test.clj#L1102
> 
> Of course I'm happy to help if you get stuck somewhere.

Great @sbelak, thanks! Do you think we need some more tests to this change?

**Comment by sbelak** [comment]
Ideally yes, as it encodes the intention and prevents regressions. The query processor is relatively tricky part of the code and flag arguments can quickly get lost in a refactor.

**Comment by amommendes** [comment]
> Ideally yes, as it encodes the intention and prevents regressions. The query processor is relatively tricky part of the code and flag arguments can quickly get lost in a refactor.

Great, you'll try add some test of the expected behavior.

**Comment by amommendes** [comment]
Hey @sbleak, WDYT about the test? In order to reduce functions/macro duplication I added sample-data/query parameter to do-with-temp-native-card, then we can test export card results with any queries.

**Review by sbelak — APPROVED** [review]
Perfect, thanks!

**Review by flamber — CHANGES_REQUESTED** [review]
This removes timezone information from other columns too, which is going to cause more problems than this tries to solve.
I really think this needs to be solved together with #10803 - and should have tests for all column types, so we don't end up with regressions like #12964.

- Sample Dataset > People > Limit 1, save question, download as CSV:

`master`
```
ID,Address,Birth Date,City,Created At,Email,Latitude,Longitude,Name,Source,State,Zip
1,9611-9809 West Rosedale Road,1986-12-12T00:00:00+01:00,Wood River,2017-10-07T01:34:35.462+02:00,borer-hudson@yahoo.com,40.71314890000001,-98.5259864,Hudson Borer,Twitter,NE,68883
```
`fix-12964`
```
ID,Address,Birth Date,City,Created At,Email,Latitude,Longitude,Name,Source,State,Zip
1,9611-9809 West Rosedale Road,1986-12-12,Wood River,2017-10-07T01:34:35.462,borer-hudson@yahoo.com,40.71314890000001,-98.5259864,Hudson Borer,Twitter,NE,68883
```

**Comment by amommendes** [comment]
Hi @flamber, thank you by fast answer and comments.
I've the following considerations regarding the points you raised: 

> This removes timezone information from other columns too, which is going to cause more problems than this tries to solve.

- I think that the changes proposed here does not remove timezone. Indeed, the current version that adds timezone in date fields when exporting saved questions, as described in the issue. It occurs because the default value for `format-rows?` is `true`. It means that date/timestamp fields will be formatted if this parameter is not passed (as in `api/card/:card-id/csv` endpoint) or not set to `false`.
Specifically in the table that you queried, the field `birth_date` is a `type/Date` field and when you exported, the query processor formatted the rows adding timezone `T00:00:00+01:00`:
image

 - Moreover, even if we desconsider these changes, the issue still remains, because we've an inconsisnt behavior between unsaved (`/api/dataset` endpoint) and saved (`api/card/:card-id/csv` endpoint) questions. That is, when exporting saved questions dates will be formatted (timezones added), while exporting unsaved questions dates will not be formatted, because this endpoint set explicitly `format-rows?` to `false`. 

> I really think this needs to be solved together with #10803
- I'm not sure if #10803 handle this issue. If I understood correctly, #10803 is dealing with Excel default columns formats. However, would be great if we could fix this issue there.

WDYT? Make sense?

**Comment by flamber** [comment]
@amommendes I don't code Clojure, I'm simply just blocking, since this would likely cause more problems for people that are expecting timezone one their non-date columns. I'll leave the resolution to smarter people than me ;-)
Ignore the Excel part, just look at my comment: https://github.com/metabase/metabase/issues/10803#issuecomment-638480542
But in any case, I really think there should be tests for date, datetime, time, etc, so we don't end up with a regression like #12964

**Comment by amommendes** [comment]
Hey @flamber, thanks!
I totally agree with tests for other DateTime family columns. 
My doubt is, should we add timezone in `date` fields? If yes, we could change the exporting process of unsaved questions to add timezone in date columns, otherwise we need to remove this formatting to saved questions.

**Review by sbelak — APPROVED** [review]
looks good

**Comment by flamber** [comment]
This is still removing timezones from datetime columns too. I'm sure this might lead to bigger problems.
I don't know what the decision is on this, but I'll leave that to Cam or Sameer.

**Comment by amommendes** [comment]
> This is still removing timezones from datetime columns too. I'm sure this might lead to bigger problems.
> I don't know what the decision is on this, but I'll leave that to Cam or Sameer.

Thanks @flamber! No problem, we can wait this issue to be handled in that broader issue (#10803). We're casting all date fields to strings as a workaround. Just for curiosity, can you provide the steps to reproduce removing timestamp from datetime fields. I can't reproduce it here and tests now is treating datetime too:

 ```clojure
  (testing "no-fomatting-datetime-with-timezone"
    (with-temp-native-card [_ card users datetime-with-timezone-query]
      (with-cards-in-readable-collection card
        (is (= ["TIMESTAMP_WITH_TIMEZONE"
                "2020-07-27T13:00:00"]
               (str/split-lines
                ((mt/user->client :rasta) :post 202 (format "card/%d/query/csv" (u/get-id card))))))))))
```

**Comment by amommendes** [comment]
> @amommendes I simply just created a Simple question > Sample Dataset > People, limit 1, save question, download CSV.
> If the question is "dirty", then you get different downloads, which I noted in [#10803 (comment)](https://github.com/metabase/metabase/issues/10803#issuecomment-638480542)

Hey @flamber, thanks!

**Comment by amommendes** [comment]
@flamber, we could close this PR, WDYT?

**Pull request closed without merging**